### PR TITLE
opam switch list-available will not display compilers flagged with avoid-version/deprecated unless --all is given

### LINKED ISSUE #6089: opam switch list-available should only display packages that aren't flagged with avoid-version/deprecated
https://discuss.ocaml.org/t/error-on-opam-install-utop-fedora/14948/6

I think that command could be improved by only returning the list of packages without those flags by default and adding a `--all` argument to get the previous behaviour

### PR DESCRIPTION
**Pull request by kit-ty-kate**
Fixes https://github.com/ocaml/opam/issues/6089

### CODE REVIEW

**Review by rjbou — CHANGES_REQUESTED** [review]
Is it possible to add a test to show the new option?

**Comment by kit-ty-kate** [comment]
> Is it possible to add a test to show the new option?

done

File: src/client/opamCommands.ml [hunk 1]
```diff
@@ -2894,6 +2907,10 @@ let switch cli =
            order = `Custom order;
         }
         compilers;
+      if not all then
+        OpamConsole.note
+          "Some compilers may have been hidden (e.g. pre-releases). \
+           If you want to display them, run: opam switch list-available --all";
```
Comment (rjbou) [inline src/client/opamCommands.ml:2913]:
```suggestion
           If you want to display them, run: 'opam switch list-available --all'";
```

File: src/client/opamCommands.ml [hunk 2]
```diff
@@ -2894,6 +2907,10 @@ let switch cli =
            order = `Custom order;
         }
         compilers;
+      if not all then
```
Comment (rjbou) [inline src/client/opamCommands.ml:2910]: It is better to show the message only if there was some hidden packages.
Ftm, we don't have the information. If we retrieve the complete `compilers` set and filter it afterwards compilers flagged with `avoid version` or `deprecated`, we can access to the diff.

File: src/client/opamCommands.ml [hunk 3]
```diff
@@ -2803,7 +2808,7 @@ let switch cli =
   let switch
       global_options build_options command print_short
       no_switch packages formula empty descr full freeze no_install deps_only repos
-      force no_action
+      force no_action all
```
[the remark this answers is not in the document]
Comment (kit-ty-kate) [reply src/client/opamCommands.ml:2812]: done

File: src/client/opamListCommand.mli [hunk 4]
```diff
@@ -57,6 +57,7 @@ type selector =
   | Pattern of pattern_selector * string
   | Atoms of atom list
   | Flag of package_flag
+  | NotFlag of package_flag
```
[the remark this answers is not in the document]
Comment (kit-ty-kate) [reply src/client/opamListCommand.mli:60]: done

File: src/client/opamCommands.ml (hunk 2 above)
Comment (kit-ty-kate) [reply src/client/opamCommands.ml]: done

File: src/client/opamCommands.ml [hunk 5]
```diff
@@ -2815,6 +2821,9 @@ let switch cli =
     if d_no_autoinstall then
       OpamConsole.warning "Option %s is deprecated, ignoring it."
         (OpamConsole.colorise `bold "--no-autoinstall");
+    if all && command <> Some `list_available then
+      OpamConsole.warning "Option %s is not used"
```
[the remark this answers is not in the document]
Comment (rjbou) [reply src/client/opamCommands.ml:2838]: I'm also wondering if we should add note about `--all` option in `list-available` section.

**Comment by rjbou** [comment]
I've updated the PR with a more complete test for witch list-available (that highlighted #6152) and squashed some commits.

File: src/client/opamCommands.ml [hunk 6]
```diff
@@ -2909,6 +2928,10 @@ let switch cli =
            order = `Custom order;
         }
         compilers;
+      if not all && not (OpamPackage.Set.equal all_compilers compilers) then
+        OpamConsole.note
+          "Some compilers may have been hidden (e.g. pre-releases). \
```
Comment (kit-ty-kate) [inline src/client/opamCommands.ml:2933]:
```suggestion
          "Some compilers have been hidden (e.g. pre-releases). \
```

**Review by rjbou — APPROVED** [review]
thanks!

**Pull request merged**